diff --git stardoc/stardoc.bzl stardoc/stardoc.bzl
index 456aca7..a065104 100644
--- stardoc/stardoc.bzl
+++ stardoc/stardoc.bzl
@@ -32,7 +32,7 @@ def _stardoc_impl(ctx):
     ])
     stardoc_args = ctx.actions.args()
     stardoc_args.add("--input=" + str(ctx.file.input.owner))
-    stardoc_args.add("--workspace_name=" + ctx.label.workspace_name)
+    stardoc_args.add("--workspace_name=" + ctx.workspace_name)
     stardoc_args.add_all(
         ctx.attr.symbol_names,
         format_each = "--symbols=%s",
