diff --git a/node_modules/@angular/compiler-cli/src/ngtsc/util/src/typescript.js b/node_modules/@angular/compiler-cli/src/ngtsc/util/src/typescript.js
index 8321ccb..fff39a5 100755
--- a/node_modules/@angular/compiler-cli/src/ngtsc/util/src/typescript.js
+++ b/node_modules/@angular/compiler-cli/src/ngtsc/util/src/typescript.js
@@ -99,6 +99,8 @@
     exports.isExported = isExported;
     function getRootDirs(host, options) {
         var rootDirs = [];
+        const cwd = host.getCurrentDirectory();
+        const fs = file_system_1.getFileSystem();
         if (options.rootDirs !== undefined) {
             rootDirs.push.apply(rootDirs, tslib_1.__spread(options.rootDirs));
         }
@@ -106,13 +108,16 @@
             rootDirs.push(options.rootDir);
         }
         else {
-            rootDirs.push(host.getCurrentDirectory());
+            rootDirs.push(cwd);
         }
         // In Windows the above might not always return posix separated paths
         // See:
         // https://github.com/Microsoft/TypeScript/blob/3f7357d37f66c842d70d835bc925ec2a873ecfec/src/compiler/sys.ts#L650
         // Also compiler options might be set via an API which doesn't normalize paths
-        return rootDirs.map(function (rootDir) { return file_system_1.absoluteFrom(rootDir); });
+        return rootDirs.map(rootDir => {
+            const rooted = fs.isRooted(rootDir) ? rootDir : fs.join(cwd, rootDir);
+            return file_system_1.absoluteFrom(rooted);
+        });
     }
     exports.getRootDirs = getRootDirs;
     function nodeDebugInfo(node) {
