diff --git a/node_modules/istanbul-api/lib/file-matcher.js b/node_modules/istanbul-api/lib/file-matcher.js
index d926363..587d5d4 100644
--- a/node_modules/istanbul-api/lib/file-matcher.js
+++ b/node_modules/istanbul-api/lib/file-matcher.js
@@ -18,8 +18,8 @@ function filesFor(options, callback) {
     let root = options.root;
     let includes = options.includes;
     let excludes = options.excludes;
-    const realpath = options.realpath;
-    const relative = options.relative;
+    const realpath = false;
+    const relative = true;
 
     root = root || process.cwd();
     includes = includes && Array.isArray(includes) ? includes : ['**/*.js'];
@@ -35,6 +35,7 @@ function filesFor(options, callback) {
             return callback(err);
         }
         if (relative) {
+            console.log("istanbul", files)
             return callback(err, files);
         }
 
@@ -62,8 +63,8 @@ function matcherFor(options, callback) {
         options = null;
     }
     options = options || {};
-    options.relative = false; //force absolute paths
-    options.realpath = true; //force real paths (to match Node.js module paths)
+    options.relative = true; //force absolute paths
+    options.realpath = false; //force real paths (to match Node.js module paths)
 
     filesFor(options, (err, files) => {
         const fileMap = Object.create(null);
