diff --git a/node_modules/jest-cli/build/cli/index.js b/node_modules/jest-cli/build/cli/index.js
index 831a476..46f7570 100644
--- a/node_modules/jest-cli/build/cli/index.js
+++ b/node_modules/jest-cli/build/cli/index.js
@@ -271,7 +271,7 @@ const getProjectListFromCLIArgs = (argv, project) => {

   if (!projects.length && process.platform === 'win32') {
     try {
-      projects.push((0, _realpathNative().sync)(process.cwd()));
+      projects.push(process.cwd());
     } catch (err) {
       // do nothing, just catch error
       // process.binding('fs').realpath can throw, e.g. on mapped drives
diff --git a/node_modules/jest-cli/build/init/index.js b/node_modules/jest-cli/build/init/index.js
index cae2fc2..7f38281 100644
--- a/node_modules/jest-cli/build/init/index.js
+++ b/node_modules/jest-cli/build/init/index.js
@@ -134,7 +134,7 @@ var _default =
   /*#__PURE__*/
   (function() {
     var _ref = _asyncToGenerator(function*(
-      rootDir = (0, _realpathNative().sync)(process.cwd())
+      rootDir = process.cwd()
     ) {
       // prerequisite checks
       const projectPackageJsonPath = _path().default.join(
