diff --git a/node_modules/jest-haste-map/build/crawlers/node.js b/node_modules/jest-haste-map/build/crawlers/node.js
index a179225..06106b9 100644
--- a/node_modules/jest-haste-map/build/crawlers/node.js
+++ b/node_modules/jest-haste-map/build/crawlers/node.js
@@ -165,7 +165,7 @@ function find(roots, extensions, ignore, callback) {
 }

 function findNative(roots, extensions, ignore, callback) {
-  const args = Array.from(roots);
+  const args = Array.from(['-L'].concat(roots));
   args.push('-type', 'f');

   if (extensions.length) {
diff --git a/node_modules/jest-haste-map/build/index.js b/node_modules/jest-haste-map/build/index.js
index eb8d84f..97ec2c1 100644
--- a/node_modules/jest-haste-map/build/index.js
+++ b/node_modules/jest-haste-map/build/index.js
@@ -403,7 +403,7 @@ class HasteMap extends _events().default {
       roots: Array.from(new Set(options.roots)),
       skipPackageJson: !!options.skipPackageJson,
       throwOnModuleCollision: !!options.throwOnModuleCollision,
-      useWatchman: options.useWatchman == null ? true : options.useWatchman,
+      useWatchman: false,
       watch: !!options.watch
     };
     this._console = options.console || global.console;
