diff --git a/node_modules/jest-runtime/build/cli/index.js b/node_modules/jest-runtime/build/cli/index.js
index 47f4f5d..ff2bbc1 100644
--- a/node_modules/jest-runtime/build/cli/index.js
+++ b/node_modules/jest-runtime/build/cli/index.js
@@ -203,7 +203,7 @@ function run(cliArgv, cliInfo) {
     return;
   }

-  const root = (0, _realpathNative().sync)(process.cwd());
+  const root = process.cwd();

   const filePath = _path().default.resolve(root, argv._[0]);

