diff --git a/go/private/repositories.bzl b/go/private/repositories.bzl
index 5649c31b..06f48769 100644
--- a/go/private/repositories.bzl
+++ b/go/private/repositories.bzl
@@ -61,7 +61,7 @@ def go_rules_dependencies():
     _maybe(
         git_repository,
         name = "org_golang_x_tools",
-        remote = "https://go.googlesource.com/tools",
+        remote = "https://github.com/golang/tools.git",
         # "latest", as of 2019-07-08
         commit = "c8855242db9c1762032abe33c2dff50de3ec9d05",
         shallow_since = "1562618051 +0000",
