diff --git a/shim.js b/shim.js
index 9ab438ec..cd2681b6 100644
--- a/shim.js
+++ b/shim.js
@@ -4,6 +4,6 @@
 
 goog.module('google3.third_party.javascript.incremental_dom.index');
 
-const incrementaldom = goog.require('incrementaldom');
+const incrementaldom = /* @type {?} */(window.incrementaldom.types);//goog.require('incrementaldom');
 
 exports = incrementaldom;
diff --git a/soyutils_velog.js b/soyutils_velog.js
index 3d63b079..792d91cf 100644
--- a/soyutils_velog.js
+++ b/soyutils_velog.js
@@ -28,7 +28,7 @@
 goog.module('soy.velog');
 goog.module.declareLegacyNamespace();
 
-const Message = goog.require('jspb.Message');
+const Message = class {};//goog.require('jspb.Message');
 const {assert} = goog.require('goog.asserts');
 const {startsWith} = goog.require('goog.string');
 
diff --git a/types.js b/types.js
index ae1ff90b..e34e7211 100644
--- a/types.js
+++ b/types.js
@@ -4,6 +4,6 @@
 
 goog.module('google3.third_party.javascript.incremental_dom.src.types');
 
-const types = goog.require('incrementaldom.src.types');
+const types = /* @type {?} */(window.incrementaldom); //goog.require('incrementaldom.src.types');
 
 exports = types;
