diff --git a/node_modules/@buxlabs/amd-to-es6/src/lib/getImportDeclaration.js b/node_modules/@buxlabs/amd-to-es6/src/lib/getImportDeclaration.js
index 1062a4d..280957f 100644
--- a/node_modules/@buxlabs/amd-to-es6/src/lib/getImportDeclaration.js
+++ b/node_modules/@buxlabs/amd-to-es6/src/lib/getImportDeclaration.js
@@ -11,7 +11,7 @@ module.exports = function (element, param) {
       })
     } else {
       specifiers.push({
-        type: 'ImportDefaultSpecifier',
+        type: 'ImportNamespaceSpecifier',
         local: {
           type: 'Identifier',
           name: param
