diff --git a/internal/common/compilation.bzl b/internal/common/compilation.bzl
index e815f37..a2b6468 100644
--- a/internal/common/compilation.bzl
+++ b/internal/common/compilation.bzl
@@ -17,7 +17,7 @@
 
 load(":common/json_marshal.bzl", "json_marshal")
 load(":common/module_mappings.bzl", "module_mappings_aspect")
-load("@build_bazel_rules_nodejs//:declaration_provider.bzl", "DeclarationInfo")
+load("@build_bazel_rules_nodejs//:providers.bzl", "DeclarationInfo")
 
 _DEBUG = False
 
