diff --git a/node_modules/jest-haste-map/build/crawlers/node.js b/node_modules/jest-haste-map/build/crawlers/node.js
index 1e7372c..3d3d5f7 100644
--- a/node_modules/jest-haste-map/build/crawlers/node.js
+++ b/node_modules/jest-haste-map/build/crawlers/node.js
@@ -217,7 +217,16 @@ function find(roots, extensions, ignore, callback) {

 function findNative(roots, extensions, ignore, callback) {
   const args = Array.from(roots);
+  // In bazel's execution environment, all source/test files are symlinked
+  // into a sandboxed directory rather than copied over directly.
+  // Jest's crawler is currently set up to use find's type flag to filter only for files.
+  // This excludes symlinked files.
+  // https://github.com/facebook/jest/pull/9351#issuecomment-780637294
+  args.push('(');
   args.push('-type', 'f');
+  args.push('-o');
+  args.push('-type', 'l');
+  args.push(')');

   if (extensions.length) {
     args.push('(');
